1か月分データを出力

takeuchiyui opened the pull request:
日付を1か月分取得する処理を追加しました。まだ未完了です。

takeuchiyui committed:
日付1か月分取得

日付を1か月分取得する処理を追加しました。まだ未完了です。

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -17,151 +14,37 @@ Private Sub TextBox1_LostFocus()
-  
     
+    'フォーマットを"2021/01"にしたいのに"2021/01/01になる"
+    targetDate = format(targetDate, "yyyy/mm")
```
日にちの部分をフォーマット時に消すことはできないのかもね。

tasobue replied:
今回の場合、どうせ1日から末日まで取得するから気にしなくていいと思う。

takeuchiyui replied:
文字列に置き換えて処理していく方法になるのでしょうか

tasobue replied:
文字列にフォーマットする必要もそもそもないか。

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -17,151 +14,37 @@ Private Sub TextBox1_LostFocus()
+    targetDate = format(targetDate, "yyyy/mm")
+    
+    '１ヵ月分取得したいから、”2021/01”が付くものをすべて取得できる？
+     targetDate2 = DateAdd("m", 1, targetDate)
```
月末の日付を取得する方法が使えます
https://www.moug.net/tech/acvba/0120017.html

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -17,151 +14,37 @@ Private Sub TextBox1_LostFocus()
+     
     'テキストの差分をとる
-    preparation = DateDiff("d", targetDate, targetDate2)
+    preparation = DateDiff("m", targetDate, targetDate2)
```
1/1から月末までの日数を取りたいので"d"のままでいいです。

takeuchiyui commented:
1か月分日付を取得

テキストボックスの値格納（変数をstring型にする）
↓
文字列の
1-4番目をyear変数に格納
2-3ｍ変数に格納
4-5ｄ変数に格納
↓
DateSerial関数を使用して１か月分取得

tasobue commented:
![image](https://user-images.githubusercontent.com/21050636/105317301-668c6180-5c05-11eb-9839-477040e84d9b.png)
DateSerialはDate型の変数を一つ返すだけだよ？
どうやって使うイメージ？？

takeuchiyui commented:
> ![image](https://user-images.githubusercontent.com/21050636/105317301-668c6180-5c05-11eb-9839-477040e84d9b.png)
> DateSerialはDate型の変数を一つ返すだけだよ？
> どうやって使うイメージ？？

引数が年月日それぞれ必要なので、テキストボックスの値を分ける必要があるかと思いました。
（訂string型ではなくdate型）

tasobue commented:
ユーザが2021/01って入力したらDateSerialで2021/01/01を作成して、あとはDateAddで残りの日付作ってくってことか。

takeuchiyui commented:
> ユーザが2021/01って入力したらDateSerialで2021/01/01を作成して、あとはDateAddで残りの日付作ってくってことか。

仕様を分かっていませんでした。
祖父江さんが送ってくださったサイトの＜詳細＞をみていて
yearやmounth,dayは具体的な数字を埋めなければいけないと思っていました。なのでテキストに入ってきたDateを分けて変数に格納しなくてはいけないと考えていました

tasobue commented:
ふーん...?
こういう手順で考えているってことであってるかな？

テキストボックスで年月(2021/01)を取得
↓
DateSerialで2021/01/01のDateを取得
↓
DateSerialで2021/01/31を取得
↓
2021/01/01 ~ 2021/01/31の日数＝31を取得
↓
日数分のループを回してファイル取得

takeuchiyui commented:
> ふーん...?
> こういう手順で考えているってことであってるかな？
> 
> テキストボックスで年月(2021/01)を取得
> ↓
> DateSerialで2021/01/01のDateを取得
> ↓
> DateSerialで2021/01/31を取得
> ↓
> 2021/01/01 ~ 2021/01/31の日数＝31を取得
> ↓
> 日数分のループを回してファイル取得

はい、そうでした

takeuchiyui committed:
1/1-1/31を取得

1/1-1/31を取得することができました。

takeuchiyui committed:
変更なし

配列を取得できたはずなのに、今までの色付け処理まで入ってくれなくなりました。
(exit subしてから次の処理に入っていない気がします)

tasobue commented:
>(exit subしてから次の処理に入っていない気がします)

確認ですが、↑ってどういう意味？
そのまま読むとexit subしちゃってるからoutputの処理は途中終了しているはずだけども。なのでoutput内でexit subの後ろに書かれているコードは実行されなくなるよ。

takeuchiyui commented:
> > (exit subしてから次の処理に入っていない気がします)
> 
> 確認ですが、↑ってどういう意味？
> そのまま読むとexit subしちゃってるからoutputの処理は途中終了しているはずだけども。なのでoutput内でexit subの後ろに書かれているコードは実行されなくなるよ。

そうですよね。そこでexit for（For文を強制終了）という書き方もあったので、試してみましたがうまくできませんでした。
ファイルがない場合Forには入らずに、次のファイルを読み込んでほしいのですが上手く書けていないです

takeuchiyui commented:
＜やること＞
１ソースをまとめる（Subを分ける）
２一致した文字の時間を出力する（現：一致したセルに色付け）
３サインインサインアウトの時間を計算して出力する

tasobue commented:
>１ソースをまとめる（Subを分ける）

まずはこれのやり方を考えるところだね。
Subの適切な名前、引数を考えてみよう。

takeuchiyui commented:
３つモジュールに分けていきます。
１output
２ファイルの内容を取得しているので”getData”
３セルへデータを書き込んでいるので”dataInsert”

tasobue commented:
個人的には動詞を先に持ってきてinsertDataの方が統一感あっていいと思う

takeuchiyui committed:
ファイル変更したため

ファイル変更のため削除いたします

takeuchiyui committed:
処理まとめ

ソースをまとめて関数呼び出し処理を追加しました

takeuchiyui committed:
getData関数

引数を受け取って戻り値を返す処理を追加しています。まだ未完了です

takeuchiyui committed:
insetData関数

未着手です。

tasobue commented on VBA/No6/module2:
```diff
@@ -0,0 +1,30 @@
+    
+    Close #1
+    
+    getData = s
```
取得したデータはreturn で返す必要があるね。
Functionの戻り値の型もFunction XX() As戻り値の型
で指定するようにしましょう。

takeuchiyui replied:
https://www.officepro.jp/excelvba/sub/index6.html
このサイトを見ていると戻り値は「プロシージャ名＝値」という書き方ではないでしょうか

tasobue replied:
VBAの戻り値の返し方はそうだったか。
ではそれで！

tasobue commented on VBA/No6/module1:
```diff
@@ -0,0 +1,47 @@
+              '戻り値を受け取る
+              s = getData(FilePath)
+                
+                Call dataInsert
```
` Call dataInsert(s)`


にしないとエラーになるんじゃない？

takeuchiyui replied:
今はcall dataInsere(s)に変更いたしました。しかしgetDataメソッド内の処理が上手くいかずに詰まっています。型が問題だと思いますが、、

tasobue replied:
> > > Byref　引数の型が一致しません
> > 
> > 
> > これがどの行で出ているか確認した？
> 
> sheet1のoutputを呼び出している処理からブレークポイントを１行ずつはり確認しましたが、outputメソッドの一番初めの行でエラーが出ています。

であれば、outputの引数あたりが怪しいですが、
ほかには、エラーが発生してデバッグで処理を止めたときカーソルが設定されている（コードの一部をドラッグで範囲選択している状態になっている）場所も確認するのがいいですね。

takeuchiyui replied:
listが範囲選択されていたので確認してみます。

takeuchiyui committed:
関数呼び出し

関数呼び出しは成功しました。
しかし戻り値が上手くいかないのと
「Byref　引数の型が一致しません」というエラー表示されています。
原因は不明です。

takeuchiyui committed:
追加処理なし

TextStream型からstring型へ変更できていない

tasobue commented:
>Byref　引数の型が一致しません

これがどの行で出ているか確認した？

tasobue commented on VBA/No6/module2:
```diff
@@ -25,6 +25,6 @@ Function getData(FilePath As String)
     Close #1
     
-    getData = s
+    getData = list
```
Function の戻り値の型の指定方法を調べて、listを返せるようにしましょう。

tasobue replied:
![image](https://user-images.githubusercontent.com/21050636/105440365-cc322980-5ca9-11eb-8539-712bc49992e3.png)
キャプチャのように戻りの型を定義してみてから動かして。
https://www.sejuku.net/blog/28904

tasobue commented:
>TextStream型からstring型へ変更できていない

これってどの行で発生しているの？

tasobue commented:
何か問題が発生しているときはどの行で発生しているかまで書いてくれるとこちらもコメントしやすいです。

takeuchiyui commented:
> > Byref　引数の型が一致しません
> 
> これがどの行で出ているか確認した？

sheet1のoutputを呼び出している処理からブレークポイントを１行ずつはり確認しましたが、outputメソッドの一番初めの行でエラーが出ています。

takeuchiyui commented:
> > TextStream型からstring型へ変更できていない
> 
> これってどの行で発生しているの？

 モジュール２     list = Split(s, vbCrLf)で起きています。
しかし今は、byrefのエラーが出ているので確認できていません。

tasobue commented:
> > > TextStream型からstring型へ変更できていない
> > 
> > 
> > これってどの行で発生しているの？
> 
> モジュール２ list = Split(s, vbCrLf)で起きています。
> しかし今は、byrefのエラーが出ているので確認できていません。

おととい原因がよくわからないエラーの調査したときみたいに、コードを部分的に消したり戻したりしながらデバッグしてエラーの原因を探ってはどうでしょうか。

takeuchiyui committed:
listの受け取り追加

Call insertData(list.ListIndex)
にして受け取るようにした。

takeuchiyui committed:
listを戻り値として返す

29行目でエラーになり、「実行エラー９１　オブジェクト変数またはWithブロックが設定されていません。」というエラーが表示されるようになった。
https://kosapi.com/post-4033/
↑上記のサイトを参考にしている

tasobue commented on VBA/No6/module2:
```diff
@@ -2,7 +2,8 @@ Function getData(FilePath As String)
     Dim fso As FileSystemObject
-    Dim s
+    Dim s As String
+    'Dim list As Object
```
listをコメントアウトしてるから宣言されなくなったために「実行エラー９１　オブジェクト変数またはWithブロックが設定されていません。」が出ているのでは」？

tasobue replied:
参考にしているサイトはリストボックスから値をとるときの話しだから今回は違うんじゃない？

takeuchiyui replied:
違うサイトを探しています。

tasobue commented on VBA/No6/module1:
```diff
@@ -30,7 +31,7 @@ Public Sub output(targetDate() As Variant)
               list = getData(FilePath)
                 
-              Call insertData(list)
+              Call insertData(list.ListIndex)
```
list.ListIndexはリストボックスとか使ったときに現在list配列のどの位置（インデックスの何番目）を取得しているかを知るためのものだから今回は違うね。
list.ListIndexで得られるのは整数値であって、一方でinsertDataに渡したいのはテキストを読み込んだ結果の配列データですよね。だからListIndexは誤りです。

takeuchiyui replied:
他の方法を探してみます

tasobue replied:
新しく作成するSubないしFunctinの引数、戻り値を決めましょう。
コードを書く前に簡単な設計をしてみるのがいいと思います。

takeuchiyui commented:
実行エラー９１が起きる原因を調べると、Setの使い方に問題があると記載があったので調査していきます

takeuchiyui committed:
listの受け取り

listの受け取り部分でエラーが発生している（３４行目）
33行目だとbyrefの引数の型が一致しません
34行目だとオブジェクトが必要
というメッセージが表示されます。
Dim　list as objectの宣言を入れて試しましたがエラーでした

takeuchiyui committed:
実行エラー９１対応

Dimの宣言を削除したことで、実行エラー９１が表示されなくなりました

tasobue commented on VBA/No6/module1:
```diff
@@ -31,7 +30,8 @@ Public Sub output(targetDate() As Variant)
                 
-              Call insertData(list.ListIndex)
+              'Call insertData(list)
+              Call insertData(s.list)
```
s. のsって何？

takeuchiyui replied:
間違っているとは思ったのですが、listという入れ物があってその中身はｓなのかと思い書いてみました

tasobue commented on VBA/No6/module1:
```diff
@@ -31,7 +30,8 @@ Public Sub output(targetDate() As Variant)
               '戻り値を受け取る
               list = getData(FilePath)
```
ローカルにソースコピーしてきて気付いたんだけど、GotHunでソース見るとlが大文字でも小文字で表示されるのね。

takeuchiyui replied:
ローカルに持ってくるとということですか。

tasobue commented on VBA/No6/module3:
```diff
@@ -0,0 +1,47 @@
+Function insertData()
```
引数が定義されていません

takeuchiyui replied:
３の関数までは入れていないので、module2と1を完了させてから書こうと考えていました

tasobue commented:
ローカルで動かしてみましたが引数と戻り値が一致しているかチェックすればゴールはすぐそこですよ！

takeuchiyui commented:
・戻り値の受け取り方を調べる
・エラー内容を調べ原因を調査する

takeuchiyui committed:
変更なし

変更ありません。

takeuchiyui committed:
戻り値の型を指定

戻り値の型を指定しました

tasobue commented on VBA/No6/module2:
```diff
@@ -1,4 +1,4 @@
-Function getData(FilePath As String)
+Function getData(FilePath As String) As String
```
Listを返しているのに戻り値が文字列なのはあっているの？エラーにならない？

takeuchiyui replied:
listの中身を見るとvariant /stringと書かれていたので両方とも試してみました。
list型を返すという意味でしょうか？

tasobue replied:
Variantでエラーになりました？
手元のソースではならなかったけど,,,

takeuchiyui replied:
variant型だと引数の型が一致しませんとエラーが出ます。ローカルに持ってこないと処理ができないとかはないですよね。

takeuchiyui committed:
引数定義

引数の定義しました。

takeuchiyui commented:
listというものを調べようと思ったのですが、リストボックスなど自分の必要な情報が今り出てきませんでした。検索内容「VBA　listとは」で行っています。
もし、おすすめのサイトなどありましたら教えて欲しいです

tasobue commented on VBA/No6/module3:
```diff
@@ -1,4 +1,4 @@
-Function insertData()
+Function insertData(log As String)
```
insertDataに渡す引数は次のどちらになりますか？
A：一つの文字列
B：文字列の配列

Aだと考えるなら引数の型はString
Bだと考えるならVariantです。

あと、logはFor Each文の中でしか有効でない変数なので怒られるんじゃないかな

takeuchiyui replied:
Bなのでvariantですね。修正します
logではなく他の変数を使用したらいいのでしょうか

tasobue replied:
受け取った引数はFor Eachで回すんでしょ？
ならFor Eachで回す対象の配列の名前と一致させとけばOKかな

tasobue commented:
> listというものを調べようと思ったのですが、リストボックスなど自分の必要な情報が今り出てきませんでした。検索内容「VBA　listとは」で行っています。
> もし、おすすめのサイトなどありましたら教えて欲しいです

VBAにList型（Listという名前でVBAがあらかじめ用意したもの）はないっぽいね。
今、竹内がList型だと思ってる変数(Listとかlistとか）は正しくはStringの変数を格納したVariant型の配列だよ。
今までの知識で扱えるものだよ。

takeuchiyui commented:
> > listというものを調べようと思ったのですが、リストボックスなど自分の必要な情報が今り出てきませんでした。検索内容「VBA　listとは」で行っています。
> > もし、おすすめのサイトなどありましたら教えて欲しいです
> 
> VBAにList型（Listという名前でVBAがあらかじめ用意したもの）はないっぽいね。
> 今、竹内がList型だと思ってる変数(Listとかlistとか）は正しくはStringの変数を格納したVariant型の配列だよ。
> 今までの知識で扱えるものだよ。

variant型でlistを宣言してあげたらいいのですね

tasobue commented:
それでうまくいくはずだよ

takeuchiyui committed:
listの宣言

listの宣言を追加しました

takeuchiyui committed:
変更なし

※質問

なぜモジュール１にはlistの宣言が必要なのに、モジュール2では宣言無くても使用できるのでしょうか。
listは変数と同じ扱いではないのですか

takeuchiyui committed:
For文変更

For文で要素を受ける変数をlistsにして、logはコレクション名なのでInの後ろに書きました

takeuchiyui commented:
1か月分データを出力することができたので、データの表示の仕方を変えていきます

tasobue commented:
> 1か月分データを出力することができたので、データの表示の仕方を変えていきます

キターーーー

takeuchiyui commented:
データ表示変更
モジュール１：一番下のセルを取得して、セルを空にする
（↑ここで空にした場合モジュール3の時にずれてしまう、、）

モジュール３：データの時間の部分だけ取得
　　　　　　　一番下の行を取得する
　　　　　　　取得したデータを出力

tasobue commented:
> ↑ここで空にした場合モジュール3の時にずれてしまう

もう少し詳しく

takeuchiyui commented:
> > ↑ここで空にした場合モジュール3の時にずれてしまう
> 
> もう少し詳しく

空にすることによって、一番下の行を取得するときに空の場所も含まれてしまうと考えました。空にしてもデータが詰めて出力されてしまう気がします

takeuchiyui commented:
空文字cells(d,場所)＝””にしたら文字として認識されるのでしょうか

tasobue commented:
空文字なら何かが入力されているセルとしては認識されないと思うけど。

takeuchiyui committed:
空文字を入力するための処理を追加

Columnsを使い列を取得しました。
型はvariantだと思ったのですが、間違っているようです。

tasobue commented:
コミットの際に書いてもらっているコメントを、私がGitHun上でコメント付けるやり方で記載してくれると
コードのどこの話しをしているかわかりやすくて助かるのでお願いします。

↓こういうコメントのしかたでお願い。
![image](https://user-images.githubusercontent.com/21050636/105463958-28aa3e80-5cd4-11eb-8297-ea993f496816.png)

tasobue commented on VBA/No6/module3:
```diff
@@ -1,47 +1,55 @@
-                
-            ElseIf InStr(lists, "LogOff") > 0 Then
+            '1番したの行を取得してセルを空にする
+            side = Cells(Rows.Count, 1).End(xlUp).Row + 1
```
Columnsはそれだけで何かを返すものじゃないよ。
Columns().XXXX みたいに後ろにColumnsの持つプロパティとかメソッド使って列に何らかの操作をしたいときに使うものです。

tasobue replied:
RangeとかCellも同じ。

takeuchiyui commented:
module1

指定の列を初期化する
↓
(For　取得した配列分回す)
配列の中身を文字列に変換
↓
フォーマットを変更
↓
パスを作成して変数に格納
↓
(If　ファイルパスが存在しない場合)
セルの一番下を取得して空文字を挿入
Cells(7 + i, "D").Value = "a"　これを使用してDFHを空にする（まだ考え中）

(else)
関数呼び出し module2

関数呼び出し module3
↓
（For終了）
↓
処理終了

takeuchiyui commented:
module3

For　取得した引数分
↓
リストの中身を参照して、時間の部分だけ抜き出す
変数に格納

↓
If　"LogOn"　"LogOff"がある場合
変数の中身をセルDFに出力（サインインサインアウト）
↓
変数に格納した値で
サインアウト-サインイン＝値をセルHに出力
↓
For終了

tasobue commented:
module1, module2って呼称だとどの処理だっけ？と思ってしまうので関数名で言ってもらっていいですか？

tasobue commented:
> 取得した配列


単に配列というのではなく、どういった情報が格納された配列かを明記してください。

takeuchiyui commented:
> module1
> 
> 指定の列を初期化する
> ↓
> (For　取得した配列分回す)
> 配列の中身を文字列に変換
> ↓
> フォーマットを変更
> ↓
> パスを作成して変数に格納
> ↓
> (If　ファイルパスが存在しない場合)
> セルの一番下を取得して空文字を挿入
> Cells(7 + i, "D").Value = "a"　これを使用してDFHを空にする（まだ考え中）
> 
> (else)
> 関数呼び出し module2
> 
> 関数呼び出し module3
> ↓
> （For終了）
> ↓
> 処理終了

tasobue commented:
> module1
> 
> 指定の列を初期化する

指定の列とは何のことでしょうか？
テキスト内容を出力する列のこと？
また、列を初期化するというと例えばA列を初期化というとA1からAの最終行までを指す、と解釈できるけどあってますか？
行・列両方指定したほうがいいのではないでしょうか？

> ↓
> (For　取得した配列分回す)
> 配列の中身を文字列に変換
> ↓
> フォーマットを変更

どういうフォーマットに変更するのか書いてください。

> ↓
> パスを作成して変数に格納
> ↓
> (If　ファイルパスが存在しない場合)
> セルの一番下を取得して空文字を挿入
> Cells(7 + i, "D").Value = "a"　これを使用してDFHを空にする（まだ考え中）

> (else)
> 関数呼び出し module2
> 
> 関数呼び出し module3
> ↓
> （For終了）
> ↓
> 処理終了

takeuchiyui commented:
> module1, module2って呼称だとどの処理だっけ？と思ってしまうので関数名で言ってもらっていいですか？

output関数

指定の列を初期化する
↓
(For　取得した配列分回す)
sheet1で取得した日付を文字列に変換
↓
フォーマットを変更
↓
パスを作成して変数に格納
↓
(If　ファイルパスが存在しない場合)
セルの一番下を取得して空文字を挿入
Cells(7 + i, "D").Value = "a"　これを使用してDFHを空にする（まだ考え中）

(else)
関数呼び出し getData

関数呼び出し insertData
↓
（For終了）
↓
処理終了

takeuchiyui commented:
> > module1, module2って呼称だとどの処理だっけ？と思ってしまうので関数名で言ってもらっていいですか？
> 
> output関数
> 
> 指定の列を初期化する
> ↓
> (For　取得した配列分回す)
> sheet1で取得した日付を文字列に変換
> ↓
> フォーマットを変更
> ↓
> パスを作成して変数に格納
> ↓
> (If　ファイルパスが存在しない場合)
> セルの一番下を取得して空文字を挿入
> Cells(7 + i, "D").Value = "a"　これを使用してDFHを空にする（まだ考え中）
> 
> (else)
> 関数呼び出し getData
> 
> 関数呼び出し insertData
> ↓
> （For終了）
> ↓
> 処理終了

module3

For　取得したパス分
↓
リストの中身を参照して、時間の部分だけ抜き出す
変数に格納

↓
If　"LogOn"　"LogOff"がある場合
変数の中身をセルDFに出力（サインインサインアウト）
↓
変数に格納した値で
サインアウト-サインイン＝値をセルHに出力
↓
For終了

tasobue commented:
> module3
> 
> For　取得した引数分

「引数として取得した配列の要素数（長さ）の分」といってください。
言葉が大雑把になる癖は設計では避けましょう。

> ↓
> リストの中身を参照して、時間の部分だけ抜き出す

「リストの中身」では曖昧なので具体的にリストの中身のデータ（例えば、ファイル名とか）に言及してください。

> 変数に格納
> 
> ↓
> If　"LogOn"　"LogOff"がある場合
> 変数の中身をセルDFに出力（サインインサインアウト）

DFといわれてもわかりにくいのでD列、F列、と記してください。

> ↓
> 変数に格納した値で
> サインアウト-サインイン＝値をセルHに出力
> ↓
> For終了

takeuchiyui commented:
> > module1
> > 指定の列を初期化する
> 
> 指定の列とは何のことでしょうか？
> テキスト内容を出力する列のこと？
> また、列を初期化するというと例えばA列を初期化というとA1からAの最終行までを指す、と解釈できるけどあってますか？
> 行・列両方指定したほうがいいのではないでしょうか？
> 
> > ↓
> > (For　取得した配列分回す)
> > 配列の中身を文字列に変換
> > ↓
> > フォーマットを変更
> 
> どういうフォーマットに変更するのか書いてください。
> 
> > ↓
> > パスを作成して変数に格納
> > ↓
> > (If　ファイルパスが存在しない場合)
> > セルの一番下を取得して空文字を挿入
> > Cells(7 + i, "D").Value = "a"　これを使用してDFHを空にする（まだ考え中）
> 
> > (else)
> > 関数呼び出し module2
> > 関数呼び出し module3
> > ↓
> > （For終了）
> > ↓
> > 処理終了

指定の列とは、今回でいうとDFH列のことです。

フォーマットは2021/01/08という文字列を20210108に変更します

tasobue commented:
ではその通りにロジックを書き直してください。

takeuchiyui commented:
> ではその通りにロジックを書き直してください。

どこの部分でしょうか？すべてロジックで書き直すということですか？

tasobue commented:
私から書いた指摘については、返答不要でそのままロジックを修正してもえらえれば大丈夫です。

tasobue commented:
> > ではその通りにロジックを書き直してください。
> 
> どこの部分でしょうか？すべてロジックで書き直すということですか？

↓に返答してくれた内容について、指摘元の箇所を直してほしいです。
指定の列とは、今回でいうとDFH列のことです。

フォーマットは2021/01/08という文字列を20210108に変更します

takeuchiyui commented:
> > > ではその通りにロジックを書き直してください。
> > 
> > 
> > どこの部分でしょうか？すべてロジックで書き直すということですか？
> 
> ↓に返答してくれた内容について、指摘元の箇所を直してほしいです。
> 指定の列とは、今回でいうとDFH列のことです。
> 
> フォーマットは2021/01/08という文字列を20210108に変更します

一部の修正ではなく、全体を修正するという認識で合っているでしょうか

takeuchiyui committed:
output関数

output関数のロジックになります

takeuchiyui committed:
insertData関数

insertData関数のロジックを書きました

tasobue commented on VBA/No6/設計書　output:
```diff
@@ -0,0 +1,24 @@
+sheet1で取得した日付を引数として受け取る
+↓
+D列初期化
```
D列のすべての行を初期化するの？
細かいけど、D列の何行目を初期化するのか具体的に書きましょう。

tasobue commented on VBA/No6/設計書　output:
```diff
@@ -0,0 +1,24 @@
+F列初期化
+H列初期化
+↓
+For　取得した日付分回す
```
「取得した日付分」では曖昧です。
具体的にお願いします。

tasobue commented on VBA/No6/設計書　output:
```diff
@@ -0,0 +1,24 @@
+↓
+ファイルパスを作成して変数(FilePath)に格納
+↓
+作成したファイルパスがない場合
```
作成したファイルパスがディレクトリ上に存在しない場合
でお願いします。

tasobue commented on VBA/No6/設計書　output:
```diff
@@ -0,0 +1,24 @@
+↓
+作成したファイルパスがない場合
+↓
+シートの一番下の行を取得して、D列F列H列に空文字を入れる
```
一番下とはどこですか？本当に一番下の行(1048576行目）ではないのはわかりますが、曖昧です。
具体的にお願いします。

tasobue commented on VBA/No6/設計書　insertData:
```diff
@@ -0,0 +1,16 @@
+For CSVファイルの行分回す
```
細かいですが、「行分」ではなく「行数分」に表現を変えてください

tasobue commented on VBA/No6/設計書　insertData:
```diff
@@ -0,0 +1,16 @@
+↓
+If "LogOn"という文字が含まれている時
+↓
+時間の部分だけ取得して変数に格納a
```
「どこから」時間の部分だけを取得しますか？
「時間の部分」だけをどうやって判定して取得しましょうか？細かく考えてみましょう。
「変数に格納a」ではなく「変数aに格納」と表現を直してください。

tasobue commented on VBA/No6/設計書　insertData:
```diff
@@ -0,0 +1,16 @@
+If "LogOn"という文字が含まれている時
+↓
+時間の部分だけ取得して変数に格納a
+D列に書き出す
```
何行目かの記載がありませんが、何行目になるかは言葉にできますか？

tasobue commented:
指摘が細かいと思うかもしれませんが、言葉にしない/できないことはコーディングできないと思って仕様を書いていってください。よろしくお願いします。

takeuchiyui committed:
設計書output

3.15.17行目を指摘内容修正いたしました。

takeuchiyui committed:
設計書insertData　指摘内容修正

3.7.12.15行目の内容を修正いたしました

tasobue commented on VBA/No6/設計書　output:
```diff
@@ -1,20 +1,20 @@
+H列7行目から1ヵ月分のセルを初期化する
 ↓
-For　取得した日付分回す
+For　Sheet1で取得した1か月分の日付を回す
```
日本語をもう少しわかりやすくしてほしいです。
あくまで回しているのは配列です。回す回数が1か月分の日付です。このあたりがごちゃまぜになった表現になっています。

tasobue commented on VBA/No6/設計書　output:
```diff
@@ -1,20 +1,20 @@
+作成したファイルパスがディレクトリ上に存在し無い場合
 ↓
-シートの一番下の行を取得して、D列F列H列に空文字を入れる
+D列の値が入っているセルの次の行を取得して、D列F列H列に空文字を入れる
```
>D列の値が入っているセルの次の行

まだ曖昧です。これだとD列2～5行に例えば値が入っていたとすると、「値が入っているセル」は2～5行のいずれでもOKです。竹内が言いたいことと一致していますか？

tasobue commented on VBA/No6/設計書　insertData:
```diff
@@ -1,16 +1,18 @@
 ↓
-時間の部分だけ取得して変数に格納a
-D列に書き出す
+csvファイルの中身からleft関数を使用して時間の部分だけ取得して変数aに格納
```
詳細設計では実際に関数を明記するのではなくleft関数でしたいことを書いてくれたらいいかな。
例えば、left関数で左から8文字をとりたいということであれば
「csvファイルの各行から左8文字を取得」という風になります。
もちろん、８文字はあくまで例ですので、時間をとるためには何文字とればいいのかは考えてください。

tasobue commented on VBA/No6/設計書　insertData:
```diff
@@ -1,16 +1,18 @@
-時間の部分だけ取得して変数に格納a
-D列に書き出す
+csvファイルの中身からleft関数を使用して時間の部分だけ取得して変数aに格納
+文字が入力されている行の次の行を取得してD列に書き出す
```
ここも「文字が入力されている次の行」では曖昧です。

tasobue commented:
ループ処理の書き方で何行目に書くのかがあいまいですが、以下のような表現を使ってわかりやすくできませんか。

（6行目から表があるとして、）
ループ処理によってセルに入力する際の開始行が６行目であると明記する。
ループのたびに入力行を下に１行ずらす。

takeuchiyui committed:
指摘内容修正

1行目１７行目を変更しました

tasobue commented on VBA/No6/設計書　output:
```diff
@@ -14,7 +14,7 @@ For　Sheet1で取得した1か月分の日付を回す
 作成したファイルパスがディレクトリ上に存在し無い場合
 ↓
-D列の値が入っているセルの次の行を取得して、D列F列H列に空文字を入れる
+7行目以降でD列に値が入っているセルの次の行を取得して、D列F列H列に空文字を入れる
```
空文字は値が入っていないセルであるとプログラムは認識します。
ですので「値が入っているセルの次の行」と記載した場合、
７行目に空文字入れたら次のループでも７行目に文字を入れることになります。
竹内の狙いでは次は８行目に入れたいのではないですか？

takeuchiyui replied:
空文字の代わりに「-」の記号を使用してもいいでしょうか

takeuchiyui committed:
指摘内容修正 insertData

ループの処理の書き始めの表現の変更しました

tasobue commented on VBA/No6/設計書　insertData:
```diff
@@ -1,16 +1,17 @@
 ※変数a 変数bは仮の名前です
 
-For CSVファイルの行数分回す
+ループ処理によってD列の7行目からCSVファイルから取得した行数分回し
```
いいたいことを詰め込みすぎてかえってわかりづらくなっているかな...

↓は個別に書いたほうが読みやすいです。
・CSVファイルから取得した行数分回す
・D列の7行目から記入
・記入行をループのたびに１行ずつずらす

tasobue commented on VBA/No6/設計書　insertData:
```diff
@@ -1,16 +1,17 @@
 ↓
 時間の部分だけ取得して変数bに格納
-文字が入力されている行の次の行を取得してF列に書き出す
+7行目以降で値が入っているセルの次の行を取得してF列に書き出す
```
ここの行の指定の仕方ですが、
上の方で「７行目から開始」、「ループごとに１行ずらす」と書いてあるので
「現在のループで参照している行」「現在行」でいいかなと思います。
実際、値が入っているかどうかではなくて、ループが何回目かで記入する行ってきまるんだよね？

takeuchiyui committed:
ループ処理の表現変更

ループ処理の表現の仕方を変更しました

takeuchiyui committed:
指摘内容修正

16行目を修正いたしました

tasobue commented on VBA/No6/設計書　insertData:
```diff
@@ -13,7 +13,7 @@ csvファイルの中身からleft関数を使用して左から10行目～14行
+現在参照している行のF列に書き出す
 ↓
 変数b-変数aをして差分を取得
 文字が入力されている行の次の行を取得してH列に書き込む
```
ここは直さなくていいですか？類似修正箇所の確認もお願いしますね。

takeuchiyui committed:
類似箇所修正

類似箇所の修正をいたしました

takeuchiyui committed:
出力処理変更

出力処理変更しました。

logonの時間取得で、1番最初のみを取得したいので変数ではなく定数にして値を固定する
logoffは1番最後の値を取得したいので変数のままループを行う

takeuchiyui committed:
ループ処理変更

ループ処理を変更いたしました
ここにどうやって位置の指定をしてinsertDataへ引数を渡すのかまだ思いつきません

takeuchiyui committed:
output　引数追加

２６行目でinsertData関数に引数を追加しました

takeuchiyui committed:
insertData 引数受け取り

引数で何を受け取りどのように使用するかを考えてみました

tasobue commented on VBA/No6/設計書　insertData:
```diff
@@ -1,4 +1,4 @@
-※定数a 変数bは仮の名前です
+outputからcsvファイルの中身（variant型）とセル位置を指定した(range型)を引数として受け取る
```
>セル位置を指定した(range型)
rangeでもありだけど、行番号をintegerで渡してinsertData側でrangeを作ればいいんじゃない？
列は固定だからinsertDataは行番号さえ知ることができればいい。

takeuchiyui replied:
> 次書くのは↓かな。
> 　・GetData関数の処理　
> 　・InsertData関数で、入力する内容の制御を考えてみる
> 　　　たとえばサインアウトされなかったときにI列のサインインの期間を入力しない
> 　　　昼休みにサインアウトし、午後からサインインした場合、I列のサインインの期間を入力しない



> いけないわけではなくて、getData関数もどんな処理をしているか言葉にできるように練習してほしい。

何か処理を足すということではなく、ロジックを丁寧に書き直すという認識でいいでしょうか

tasobue commented:
次書くのは↓かな。
　・GetData関数の処理　
　・InsertData関数で、入力する内容の制御を考えてみる
　　　たとえばサインアウトされなかったときにI列のサインインの期間を入力しない
　　　昼休みにサインアウトし、午後からサインインした場合、I列のサインインの期間を入力しない

takeuchiyui commented:
> 次書くのは↓かな。
> 　・GetData関数の処理　
> 　・InsertData関数で、入力する内容の制御を考えてみる
> 　　　たとえばサインアウトされなかったときにI列のサインインの期間を入力しない
> 　　　昼休みにサインアウトし、午後からサインインした場合、I列のサインインの期間を入力しない

getData関数は今のままでは何がいけないのでしょうか？中身を読み込む処理はできていると思います。

tasobue commented:
いけないわけではなくて、getData関数もどんな処理をしているか言葉にできるように練習してほしい。

takeuchiyui commented:
> 次書くのは↓かな。
> 　・GetData関数の処理　
> 　・InsertData関数で、入力する内容の制御を考えてみる
> 　　　たとえばサインアウトされなかったときにI列のサインインの期間を入力しない
> 　　　昼休みにサインアウトし、午後からサインインした場合、I列のサインインの期間を入力しない

昼休みにサインアウトして午後からサインインした場合でも、変数・定数に格納しているのは最初にサインインした時間と最後にサインアウトした時間なので計算できるのではないでしょうか

takeuchiyui committed:
引数受け取り

追加内容

引数受け取りで行番号を受け取る
サインアウトされなかったときにI列のサインインの期間を入力しない

tasobue commented:
サンプルのExcelのレイアウトみると「昼にサインアウト、午後にサインイン」した場合にI列に値入れてないんですよね。

takeuchiyui commented:
> サンプルのExcelのレイアウトみると「昼にサインアウト、午後にサインイン」した場合にI列に値入れてないんですよね。

一番最初のサインインと一番最後のサインアウトの時間だけを取得したいいと認識しているのですが合っているでしょうか。
もし合っている場合、今のロジックの書き方的に最初のサインインと最後のサインアウトを取得するような処理になっていると思います。サインアウトしていない場合はI列に値を入れないという処理を書くということでしょうか

takeuchiyui committed:
設計書　getData

getDataの処理の流れを書きました

tasobue commented:
>一番最初のサインインと一番最後のサインアウトの時間だけを取得したいいと認識しているのですが合っているでしょうか。

あっています。

>今のロジックの書き方的に最初のサインインと最後のサインアウトを取得するような処理になっていると思います。

同意です。

>サインアウトしていない場合はI列に値を入れないという処理を書くということでしょうか

あっています。

上記に加えて、「昼にサインアウト、午後にサインイン」した場合にI列に値を入れないというパターンも実装する必要があるといっています。

takeuchiyui commented:
> > 一番最初のサインインと一番最後のサインアウトの時間だけを取得したいいと認識しているのですが合っているでしょうか。
> 
> あっています。
> 
> > 今のロジックの書き方的に最初のサインインと最後のサインアウトを取得するような処理になっていると思います。
> 
> 同意です。
> 
> > サインアウトしていない場合はI列に値を入れないという処理を書くということでしょうか
> 
> あっています。
> 
> 上記に加えて、「昼にサインアウト、午後にサインイン」した場合にI列に値を入れないというパターンも実装する必要があるといっています。

サインイン9：00
サインアウト12：00
サインイン13：00
サインアウト17：00
という場合はサインインとサインアウトを時間を出さない。

サインイン9：00
サインアウト17：00
という場合のみ時間を取得するということですね

tasobue commented:
もうしわけない。柴田さんに仕様を確認したところ、
「「昼にサインアウト、午後にサインイン」した場合もI列に値を**入れる**」が正しいようです。
もらった資料がそもそも間違っているそうです。

takeuchiyui committed:
条件文追加

「昼にサインアウト、午後にサインイン」した場合にI列を空文字を入れる処理をついかしました

tasobue commented on VBA/No6/設計書　insertData:
```diff
@@ -3,19 +3,25 @@ outputからcsvファイルの中身（variant型）とセル位置を指定し
 CSVファイルから取得した行数分回す
 ↓
-If "LogOn"という文字が含まれている時
+If "LogOn"という文字が1つ含まれている時
```
確認させて。「一つ含まれている」というように数を指定したのはなぜ？

tasobue replied:
16行目のLogOffについても「一つ含まれている」としたのはなぜ？

takeuchiyui replied:
サインイン9：00
サインアウト12：00
サインイン13：00
サインアウト17：00
という場合はサインインとサインアウトを時間を出さない。

この仕様で条件分岐を書こうと思うと、条件をlogonという文字が1つの場合の時のみI列に時間を表示させたかったためこのような書き方をしました

takeuchiyui commented:
> もうしわけない。柴田さんに仕様を確認したところ、
> 「「昼にサインアウト、午後にサインイン」した場合もI列に値を**入れる**」が正しいようです。
> もらった資料がそもそも間違っているそうです。

サインイン9：00
サインアウト12：00
-----------------------------ここで時間計算１
サインイン13：00
サインアウト17：00
---------------------------ここで時間計算２

１と２の時間を加算してI列に表示したらいいのでしょうか

tasobue commented:
Excelの資料だと休憩時間は稼働時間に加算してるか確認できる？

takeuchiyui commented:
> Excelの資料だと休憩時間は稼働時間に加算してるか確認できる？

すみません。Excelファイルを確認しようとしたのですが、Excel資料はどれのことを指しているのでしょうか。
以前祖父江さんにキャプチャで送っていただいたExcelのことでしょうか？
それともトヨタ産業記念館のフォルダのどこかにあるExcelファイルのことでしょうか。

tasobue commented:
Chatに送った

takeuchiyui committed:
指摘内容修正

サインインとサインアウトの時間計算はExcel関数で行うため削除しました

tasobue commented:
だいぶ言葉で処理を書き起こせてきましたね。
じゃあそろそろコードを書いていきましょう。
設計書に書いた処理をなぞりながらコーディングしてってください。
実際に書くと見落としていた部分がでてくるとは思いますがそこはバグをつぶしながら完成を目指してください。

takeuchiyui committed:
output関数　改修版

ロジックに沿って処理を書きました

takeuchiyui committed:
insertData関数　改修版

ロジックに沿って書きました
